content/style/go.md

```diff
@@ -1,3 +1,1659 @@
+
+* Complex expressions that cannot be represented as single assignments.
+* Pluggable hooks, such as `database/sql` dialects, encoding type registries, etc.
+
+### Exit in Main
+
+Go programs use [`os.Exit`] or [`log.Fatal*`] to exit immediately. (Panicking is not a good way to exit programs, please don't panic.)
+
+[`os.Exit`]: https://golang.org/pkg/os/#Exit
+[`log.Fatal*`]: https://golang.org/pkg/log/#Fatal
+
+Call one of `os.Exit` or `log.Fatal*` **only in `main()`**. All other functions should return errors to signal failure.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+func main() {
+  body := readFile(path)
+  fmt.Println(body)
+}
+
+func readFile(path string) string {
+  f, err := os.Open(path)
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  b, err := ioutil.ReadAll(f)
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  return string(b)
+}
+```
+
+</td><td>
+
+```go
+func main() {
+  body, err := readFile(path)
+  if err != nil {
+    log.Fatal(err)
+  }
+  fmt.Println(body)
+}
+
+func readFile(path string) (string, error) {
+  f, err := os.Open(path)
+  if err != nil {
+    return "", err
+  }
+
+  b, err := ioutil.ReadAll(f)
+  if err != nil {
+    return "", err
+  }
+
+  return string(b), nil
+}
+```
+
+</td></tr>
+</tbody></table>
+
+Rationale: Programs with multiple functions that exit present a few issues:
+
+* Non-obvious control flow: Any function can exit the program, so it becomes difficult to reason about the control flow.
+* Difficult to test: A function that exits the program will also exit the test calling it.
+  This makes the function difficult to test and introduces risk of skipping other tests that have not yet been run by `go test`.
+* Skipped cleanup: When a function exits the program, it skips function calls enqueued with `defer` statements.
+  This adds risk of skipping important cleanup tasks.
+
+#### Exit Once
+
+If possible, prefer to call `os.Exit` or `log.Fatal` **at most once** in your `main()`.
+If there are multiple error scenarios that halt program execution, put that logic under a separate function and return errors from it.
+
+This has the effect of shortening your `main()` function and putting all key business logic into a separate, testable function.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+package main
+
+func main() {
+  args := os.Args[1:]
+  if len(args) != 1 {
+    log.Fatal("missing file")
+  }
+  name := args[0]
+
+  f, err := os.Open(name)
+  if err != nil {
+    log.Fatal(err)
+  }
+  defer f.Close()
+
+  // If we call log.Fatal after this line,
+  // f.Close will not be called.
+
+  b, err := ioutil.ReadAll(f)
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  // ...
+}
+```
+
+</td><td>
+
+```go
+package main
+
+func main() {
+  if err := run(); err != nil {
+    log.Fatal(err)
+  }
+}
+
+func run() error {
```

Comment:
Since we have a pretty established pattern with a `run()` function that returns a numeric status code, I think it may be a better example here.

Reply:
True, I vastly prefer the `error` way personally rather than what we do though. Maybe it's smth we can discuss. What are your thoughts about this? @Maelkum @ricardomgoncalves @awfm9 

For me what's neat with this is that instead of having a `log.Error()....` followed by a `return success`/`return failure` in each case, we just return an error like in any other function and it is logged properly at the `main` level. If `nil` is returned, nothing is logged and the program exits successfully.

I feel like this way is both more idiomatic and clearer/more readable.

Reply:
That's true, I think the function returning an error is more idiomatic. 👍

Reply:
I also think returning an error is more idiomatic and readable.

Reply:
I'll open an issue for making this change in our repos then.